x/farming/types/proposal.go

```diff
@@ -43,65 +40,90 @@ func (p *PublicPlanProposal) ProposalRoute() string { return RouterKey }
 func (p *PublicPlanProposal) ProposalType() string { return ProposalTypePublicPlan }
 
 func (p *PublicPlanProposal) ValidateBasic() error {
-	for _, plan := range p.Plans {
-		_, ok := plan.GetCachedValue().(PlanI)
-		if !ok {
-			return fmt.Errorf("expected planI")
+	if p.AddRequestProposals == nil && p.UpdateRequestProposals == nil && p.DeleteRequestProposals == nil {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "proposal request must not be empty")
+	}
+
+	for _, ap := range p.AddRequestProposals {
+		if err := ap.Validate(); err != nil {
+			return err
+		}
+	}
+
+	for _, up := range p.UpdateRequestProposals {
+		if err := up.Validate(); err != nil {
+			return err
+		}
+	}
+
+	for _, dp := range p.DeleteRequestProposals {
+		if err := dp.Validate(); err != nil {
+			return err
 		}
-		// TODO: PlanI needs ValidateBasic()?
-		// if err := p.ValidateBasic(); err != nil {
-		// 	return err
-		// }
 	}
 	return gov.ValidateAbstract(p)
 }
 
 func (p PublicPlanProposal) String() string {
-	return fmt.Sprintf(`Create FixedAmountPlan Proposal:
-  Title:       %s
-  Description: %s
-  Plans: 	   %s
-`, p.Title, p.Description, p.Plans)
+	return fmt.Sprintf(`Public Plan Proposal:
+  Title:       			  %s
+  Description: 		      %s
+  AddRequestProposals: 	  %s
+  UpdateRequestProposals: %s
+  DeleteRequestProposals: %s
+`, p.Title, p.Description, p.AddRequestProposals, p.UpdateRequestProposals, p.DeleteRequestProposals)
 }
 
-// PackPlans converts PlanIs to Any slice.
-func PackPlans(plans []PlanI) ([]*types.Any, error) {
-	plansAny := make([]*types.Any, len(plans))
-	for i, plan := range plans {
-		msg, ok := plan.(proto.Message)
-		if !ok {
-			return nil, fmt.Errorf("cannot proto marshal %T", plan)
-		}
-		any, err := types.NewAnyWithValue(msg)
-		if err != nil {
-			return nil, err
-		}
-		plansAny[i] = any
+func (p *AddRequestProposal) Validate() error {
+	if len(p.Name) > MaxNameLength {
+		return sdkerrors.Wrapf(ErrInvalidNameLength, "plan name cannot be longer than max length of %d", MaxNameLength)
 	}
-
-	return plansAny, nil
+	if _, err := sdk.AccAddressFromBech32(p.FarmingPoolAddress); err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid farming pool address %q: %v", p.FarmingPoolAddress, err)
+	}
+	if _, err := sdk.AccAddressFromBech32(p.TerminationAddress); err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid termination address %q: %v", p.TerminationAddress, err)
+	}
+	if err := p.StakingCoinWeights.Validate(); err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid staking coin weights: %v", err)
+	}
+	if !p.EndTime.After(p.StartTime) {
+		return sdkerrors.Wrapf(ErrInvalidPlanEndTime, "end time %s must be greater than start time %s", p.EndTime, p.StartTime)
+	}
+	if !p.EpochAmount.Empty() && !p.EpochRatio.IsZero() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "epoch amount or epoch ratio must be provided")
```

Comment:
It checks if both `EpochAmount` *and* `EpochRatio` has been set, but don't check if those two are both empty.
We need another if-statement for that.

Reply:
You're right. I'll add another if-statement for that.